views/mdc/assets/js/utils/date-parser.js

```diff
@@ -0,0 +1,75 @@
+const UNICODE_NUMERAL = /\p{N}/giu;
+
+export class DateParser {
+    constructor(formatter = new Intl.DateTimeFormat(navigator.languages)) {
+        this.formatter = formatter;
+    }
+
+    parse(dateString) {
+        const parts = this.formatter.formatToParts();
+        const codePoints = Array.from(dateString);
+        const accumulator = {
+            day: [],
+            month: [],
+            year: [],
+        };
+        let c = codePoints.shift();
+
+        for (const {type} of parts) {
+            switch (type) {
+            case 'day':
+            case 'month':
+            case 'year':
+            case 'relatedYear':
+                // capture a span of unicode numerals:
+                while (c && c.match(UNICODE_NUMERAL)) {
+                    accumulator[type].push(transliterateNumber(c));
+                    c = codePoints.shift();
+                }
+                break;
+            default:
+                // consume a span of non-numeric code points:
+                while (c != null && !c.match(UNICODE_NUMERAL)) {
+                    c = codePoints.shift();
+                }
+                break;
+            }
+        }
+
+        const day = parseInt(accumulator.day.join(''));
+        const month = parseInt(accumulator.month.join(''));
+        const year = parseInt(accumulator.year.join(''));
+
+        if (isNaN(day) || isNaN(month) || isNaN(year)) {
+            console.warn('Invalid date:', dateString);
+            return null;
+        }
+
+        const monthIndex = month - 1; // months in JS start at 0!
+        const date = new Date(year, monthIndex, day);
+
+        if (isNaN(date.getTime())) {
+            // day, month, and year may be valid integers (as per parseInt), but they may not
+            // collectively make a valid date.
+            console.warn('Invalid date:', dateString);
+            return null;
+        }
+
+        console.debug('DateTime: parsed', dateString, 'as', date);
+
+        return date.toISOString().substring(0, 10); // just date, without time and zone components
+    }
+}
+
+/**
+ * Transliterates a numeral code point to an (Western) Arabic number.
+ *
+ * Western Arabic, Eastern Arabic (Indo-Arabic), Persian, and Urdu numerals are supported (though
+ * Western Arabic numerals are returned as-is, of course).
+ * Abjad numerals are not supported, as there is no value for zero in the Abjad numeral system.
+ * @param {String} codePoint The numeral code point to transliterate
+ * @return {String}
+ */
+function transliterateNumber(codePoint) {
+    return codePoint.replace(/[\u0660-\u0669\u06f0-\u06f9]/g, (s) => s.charCodeAt(0) & 0xf);
+}
```

Comment:
Should we be writing our own date parser? There are already libraries to do this, like moment.js or date.js.

Reply:
we don't have a fixed date format string to pass to these date libraries like they expect. we can attempt to build one, but I'm not confident in saying that it's any more or less reliable than consuming tokens like this parser is doing. I don't know that it's worth an extra 2+ KB just to capture spans of digits like this parser is doing.

this parser is simple and matches a small string against a list of tokens provided by the browser. what issues specifically do you see with it?

Reply:
maybe something like this?
```javascript
const parts = new Intl.DateTimeFormat(["en-GB"]).formatToParts();
const format = parts.map((p) => {
  switch (p.type) {
  case "month":
    return "MM";
  case "day":
    return "DD";
  case "year":
    return "YYYY";
  case "literal":
    return p.value;
  }
}).join("")
// "DD/MM/YYYY"
moment("31/10/2020", format)
// "2020-10-31T00:00:00-04:00"
```

but I don't know that this really works for any locale it may come across. do you have any suggestions?

Reply:
I'm not sure what you mean by "a fixed date format string". Day.js is capable of parsing dates from strings. I tried it with a couple of different date formats and it seems reliable. I would trust a popular library like this to correctly parse dates over anything we can write ourselves.

```javascript
dayjs("April 12, 2024").toISOString()
```

Reply:
day.js's and moment's docs both state that the string-parsing constructor only supports one or two known formats consistently: ISO 8601 and RFC 2822. anything other than those formats fall back to the browser's `Date` constructor, which of course varies from browser to browser.

day.js[^2]:

>Warning: Browser support for parsing strings [is inconsistent](http://dygraphs.com/date-formats.html). Because there is no specification on which formats should be supported, what works in some browsers will not work in other browsers.
>
>For consistent results parsing anything other than ISO 8601 strings, you should use [String + Format](https://momentjs.com/docs/#/parsing/string-format/).

momentjs[^1]:

> For consistent results parsing anything other than ISO 8601 strings, you should use [String + Format](https://day.js.org/docs/en/parse/string-format).

so any successes you're seeing with strings that aren't those formats isn't reliably. ergo, we need a fixed string format to pass to these libraries.

---
[^1]: https://day.js.org/docs/en/parse/string
[^2]: https://momentjs.com/docs/#/parsing/string